File: cs50/week6/sentimental-credit/credit.py
# promt the user for a card numb
from cs50 import get_int


def main():
    cn2 = 0
    while True:
        n = get_int("Number: ")
        strn = str(n)
        ln = len(strn)
        if ln >= 1 and ln <= 16:
            break
        # print("INVALID")

    fn = int(strn[0])
    len2 = ln - 2
    len1 = ln - 1
    if (fn >= 3 and fn <= 5) and (ln == 13 or ln == 15 or ln == 16):
        while len2 >= 0:
            cn = int(strn[len2]) * 2
            if cn >= 10:
                cn -= 9
            cn2 += cn
            len2 -= 2
        # print(cn2)
        while len1 >= 0:
            cn2 += int(strn[len1])
            len1 -= 2
        # print(cn2)
        if cn2 % 10 == 0:
            which_card(strn)
        else:
            print("INVALID")
    else:
        print("INVALID")


def which_card(strn):
    n = int(strn[:2])
    if n == 34 or n == n == 37:
        print("AMEX")
    elif n >= 51 and n <= 55:
        print("MASTERCARD")
    elif int(strn[:1]) == 4:
        print("VISA")
    else:
        print(n / 10)
        print("INVALID")


main()

File: cs50/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <stdlib.h>

long prompt_card(void);
void check_sum(long n, int sn);
int count_numb(long n);
int start_numb(long n);
void which_card(long n);

int main(void)
{
    long   n;
    int   nb;
    int   sn;
    //prompt a card nub
    n = prompt_card();
    //count numb
    nb = count_numb(n);
    sn = start_numb(n);
    //calculate the checksum
    if ((nb == 13 || nb == 15 || nb == 16) && (sn >= 3 && sn <= 5))
        check_sum(n, sn);
    else
    {
       return(printf("INVALID\n"), 0);
    }
    which_card(n);
}

long prompt_card(void)
{
    long n;
    do
    {
        n = get_long("card numb? ");
    }
    while (n <= 0 || n >= 10000000000000000);
    return (n);
}

void check_sum(long n, int sn)
{
    long x = 10;
    long y1 = 0;
    long y2 = 0;
    long y3 = 0;
    long y4 = 0;
    long m  = 0;

    m = n;
    while (m / x != 0)
    {
        m = m / x;
        y1 = (m % 10) * 2;
        if (y1 >= 10)
            y1 = y1 - 9;
        y2 = y2 + y1;
        x = 100;
    }
    m = n;
    x = 1;
    while (m / x != 0)
    {
        m = m / x;
        y3 = (m % 10);
        y4 = y4 + y3;
        x = 100;
    }
    if ((y4 + y2) % 10 != 0)
    {
        printf("INVALID\n");
        exit(0);
    }
}

int start2_numb(long n)
{
    int i = 1;
    while (n >= 100)
    {
        i++;
        n = n / 10;
    }
    return (n);
}
int start_numb(long n)
{
    int i = 1;
    while (n / 10 != 0)
    {
        i++;
        n = n / 10;
    }
    return (n);
}

int count_numb(long n)
{
    int i = 1;
    while (n / 10 != 0)
    {
        i++;
        n = n / 10;
    }
    return (i);
}

void which_card(long n)
{
    int s2n;

    s2n = start2_numb(n);
    if (s2n == 34 || s2n == 37)
        printf("AMEX\n");
   else if (s2n >= 51 && s2n <= 55)
        printf("MASTERCARD\n");
    else if (s2n / 10 == 4)
        printf("VISA\n");
    else
         printf("INVALID\n");
}